File: 102_binary_tree_level_order_traversal.rb
=begin
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]
=end

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[][]}
def level_order(root)
    @levels = []
    if root.nil?
        return @levels
    end
    
    helper(root, 0)
    
    return @levels
end

def helper(node, level)
    if @levels.length == level
        @levels.push([])
    end
    
    @levels[level] << node.val
    
    if !node.left.nil?
        helper(node.left, level + 1)
    end
    if !node.right.nil?
        helper(node.right, level + 1)
    end
end

File: 102_binary_tree_level_order_traversal.go
/*
Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

 

Example 1:


Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]
Example 2:

Input: root = [1]
Output: [[1]]
Example 3:

Input: root = []
Output: []
 

Constraints:

The number of nodes in the tree is in the range [0, 2000].
-1000 <= Node.val <= 1000
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    result := [][]int{}
    queue := []*TreeNode{root}
    for len(queue) != 0 {
        level := []int{}
        l := len(queue)
        for i := 0; i < l; i++ {
            if queue[0] != nil {
                level = append(level, queue[0].Val)
                queue = append(queue, queue[0].Left)
                queue = append(queue, queue[0].Right)
            }   
            queue = queue[1:]
        }
        result = append(result, level)
    }
    return result[:len(result)-1]
}

